compiler/src/iree/compiler/Codegen/Common/GPU/GPUConvertToCoalescedDMA.cpp

```diff
@@ -0,0 +1,168 @@
+#include "llvm/Support/Debug.h"
+#include "mlir/Dialect/Arith/IR/Arith.h"
+#include "mlir/Dialect/GPU/IR/GPUDialect.h"
+#include "mlir/Dialect/SCF/IR/SCF.h"
+#include "mlir/Dialect/Tensor/IR/Tensor.h"
+#include "mlir/IR/Builders.h"
+#include "mlir/IR/BuiltinOps.h"
+#include "mlir/IR/PatternMatch.h"
+#include "mlir/Pass/Pass.h"
+#include "mlir/Transforms/GreedyPatternRewriteDriver.h"
+
+#define DEBUG_TYPE "iree-codegen-gpu-convert-to-coalesced-dma"
+
+namespace mlir::iree_compiler {
+
+#define GEN_PASS_DEF_GPUCONVERTTOCOALESCEDDMAPASS
+#include "iree/compiler/Codegen/Common/GPU/Passes.h.inc"
+
+namespace {
+
+// Tiles linalg_ext.gather ops according to the subgroup tiling level in the
+// lowering config and converts them to scf.forall with coalesced_gather_dma.
+//
+// Before:
+//   %result = linalg_ext.gather
+//       ins(%source, %indices : tensor<128xf32>, tensor<64xi32>)
+//       outs(%init : tensor<64xf32>)
+//
+// After (with subgroup tile size [32]):
+//   %result = scf.forall (%iv) in (64) step (32)
+//       shared_outs(%shared = %init) -> (tensor<64xf32>) {
+//     %indices_slice = tensor.extract_slice %indices[%iv] [32] [1]
+//     %indices_idx = arith.index_cast %indices_slice : tensor<32xi32> to
+//     tensor<32xindex> scf.forall.in_parallel {
+//       %dma = iree_gpu.coalesced_gather_dma %indices_idx, %source, %shared
+//     }
+//   } {mapping = [#gpu.warp<linear_dim_0>]}
+//
+struct ConverGatherOpToCoalescedDMA
+    : public OpRewritePattern<IREE::LinalgExt::GatherOp> {
+  using OpRewritePattern<IREE::LinalgExt::GatherOp>::OpRewritePattern;
+
+  LogicalResult matchAndRewrite(IREE::LinalgExt::GatherOp gatherOp,
+                                PatternRewriter &rewriter) const override {
+    // The GatherOp must be marked with a lowering config.
+    auto loweringConfig =
+        getLoweringConfig<IREE::GPU::LoweringConfigAttr>(gatherOp);
+    if (!loweringConfig) {
+      return failure();
+    }
+
+    auto subgroupTileSizes = loweringConfig.getTilingLevelSizes(
+        rewriter, llvm::to_underlying(IREE::GPU::TilingLevel::Subgroup),
+        gatherOp);
+    if (subgroupTileSizes.empty()) {
+      return failure();
+    }
+
+    // Actual tiling to CoalescedDMA:
+    Value source = gatherOp.getInputs()[0];
+    Value indices = gatherOp.getInputs()[1];
+    Value init = gatherOp.getOutputs()[0];
+
+    auto initType = cast<RankedTensorType>(init.getType());
+    int64_t rank = initType.getRank();
+
+    SmallVector<OpFoldResult> tileSizes;
+    SmallVector<OpFoldResult> lowerBounds;
+    SmallVector<OpFoldResult> upperBounds;
+    SmallVector<OpFoldResult> steps;
+
+    Location loc = gatherOp.getLoc();
+    for (int64_t i = 0; i < rank; ++i) {
+      OpFoldResult tileSize = subgroupTileSizes[i];
+      tileSizes.push_back(tileSize);
+      lowerBounds.push_back(rewriter.getIndexAttr(0));
+      upperBounds.push_back(rewriter.getIndexAttr(initType.getDimSize(i)));
+      steps.push_back(tileSize);
```

Comment:
We need either verification that the tile sizes here divide the copy size evenly or to emit the appropriate `affine.min`s to account for boundary tiles. Ideally we would do the latter for completeness if you're willing to add that here.

Reply:
I added verification on the division for now. I will think about the boundary tiles as it might impact the actual lowering..